onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -359,6 +373,33 @@ Status CUDAExecutionProvider::SetComputeStream(void* stream) {
   return Status::OK();
 }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+void CUDAExecutionProvider::CaptureBegin()  {
+  ORT_THROW_IF_ERROR(Sync());
+  graph_.Reset();
+  graph_.CaptureBegin();
+}
+
+void CUDAExecutionProvider::CaptureEnd() {
+  graph_.CaptureEnd();
+  ORT_THROW_IF_ERROR(Sync());
+  is_graph_captured_ = true;
+}
+
+bool CUDAExecutionProvider::ConfiguredForGraphCapture() {
+  return info_.enable_cuda_graph;
+}
+
+bool CUDAExecutionProvider::IsGraphCaptured() {
+  return is_graph_captured_;
+}
+
+Status CUDAExecutionProvider::GraphReplay() {
+  ORT_RETURN_IF_ERROR(graph_.Replay());
+  return Status::OK();
```

Comment:
Not understanding how we can do without the stream sync offered by OnRunEnd(true) here ?

Reply:
I realized that `cudaGraphLaunch()` run in synchronous form instead of async. Here is the profiling result  of running 10 times of `Replay()`:
```
for i in range(10):
    session.run_with_iobinding(io_binding)
``` 

<img width="1117" alt="image" src="https://user-images.githubusercontent.com/5057740/152896340-6b41a8b9-94a6-46d1-81d4-5cb9593896b3.png">

Reply:
Looking at the Nvidia docs, I don't see how cudaGraphLaunch() can be synchronous - https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g1accfe1da0c605a577c22d9751a09597.

Even the samples from Nvidia use a stream sync after cudaGraphLaunch() - 

![image](https://user-images.githubusercontent.com/9969784/152913690-0e154e22-3979-47aa-89f3-52cfa8483230.png)

Reply:
You are correct. The above model is too small. I used a large model to profile and saw `cudaGraphLaunch()` was launched asynchronously. I will update the PR tomorrow. 

While testing the large model, I found that a few warmups are needed for some models. I will try to fix this issue as well tomorrow.

Reply:
Warmup is supported by adding `cuda_graph_warmup_runs` into ep options. Also `cudaStreamSynchronize` is added to `Status CUDAGraph::Replay()`.